sys/psa_crypto/psa_hmac.c

```diff
@@ -0,0 +1,248 @@
+                              uint8_t *mac,
+                              size_t mac_size,
+                              size_t *mac_length)
+{
+    psa_mac_operation_t operation = psa_mac_operation_init();
+    psa_status_t status;
+
+    status = psa_mac_sign_setup(&operation, key, alg);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    status = psa_mac_update(&operation, input, input_length);
+    if (status != PSA_SUCCESS) {
+        psa_mac_abort(&operation);
+        return status;
+    }
+
+    status = psa_mac_sign_finish(&operation, mac, mac_size, mac_length);
+    if (status != PSA_SUCCESS) {
+        psa_mac_abort(&operation);
+    }
+
+    return status;
+}
+
+psa_status_t psa_hmac_verify(psa_key_id_t key,
+                             psa_algorithm_t alg,
+                             const uint8_t *input,
+                             size_t input_length,
+                             const uint8_t *mac,
+                             size_t mac_length)
+{
+    psa_mac_operation_t operation = psa_mac_operation_init();
+    psa_status_t status;
+
+    status = psa_mac_verify_setup(&operation, key, alg);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    status = psa_mac_update(&operation, input, input_length);
+    if (status != PSA_SUCCESS) {
+        psa_mac_abort(&operation);
+        return status;
+    }
+
+    status = psa_mac_verify_finish(&operation, mac, mac_length);
+    if (status != PSA_SUCCESS) {
+        psa_mac_abort(&operation);
+    }
+
+    return status;
+}
+
+psa_status_t psa_hmac_setup(psa_mac_operation_t *operation,
+                            const uint8_t *key_buffer,
+                            size_t key_buffer_size,
+                            psa_algorithm_t alg)
+{
+    psa_algorithm_t hash_alg = PSA_ALG_GET_HASH(alg);
+    psa_status_t status;
+    size_t block_length;
+
+    block_length = PSA_HASH_BLOCK_LENGTH(hash_alg);
+    if (!block_length) {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    operation->alg = alg;
+    status = psa_hmac_setup_key(operation, key_buffer, key_buffer_size);
+    if (status != PSA_SUCCESS) {
+        // Clear key data
+        explicit_bzero(&operation->block, block_length);
+        return status;
+    }
+
+    status = psa_hash_setup(&operation->hash, hash_alg);
+    if (status != PSA_SUCCESS) {
+        // Clear key data
+        explicit_bzero(&operation->block, block_length);
+        return status;
+    }
+
+    for (size_t i = 0; i < block_length; i++) {
+        operation->block[i] = operation->block[i] ^ 0x36;
+    }
+
+    status = psa_hash_update(&operation->hash, operation->block, block_length);
+    if (status != PSA_SUCCESS) {
+        psa_hash_abort(&operation->hash);
+        // Clear key data
+        explicit_bzero(&operation->block, block_length);
+        return status;
+    }
+
+    for (size_t i = 0; i < block_length; i++) {
+        operation->block[i] = operation->block[i] ^ 0x36;
+    }
```

Comment:
why do you do this?

Reply:
I do this to undo the previous XOR operation, so that i can reuse the buffer content, i will add a comment explaining this.